sktime/forecasting/compose/_category_compositor.py

```diff
@@ -0,0 +1,479 @@
+
+
+class CategoryCompositor(BaseForecaster):
+    """Compositor that utilizes varying forecasters by time series data's nature.
+
+    Applies a series-to-primitives transformer on a given time series. Based on the
+    generated value from the transformer, one of multiple forecasters provided by
+    the user in the form of a dictionary (key => category, value => forecaster) is
+    selected. Finally, the chosen forecaster is fit to the data for future predictions.
+
+    Parameters
+    ----------
+    forecasters : dict[sktime forecasters]
+        dict of forecasters with the key corresponding to categories generated
+        by the given transformer and the value corresponding to a sktime forecaster.
+
+    transformer : sktime transformer, default = ADICVTransformer()
+        A series-to-primitives sk-time transformer that generates a value
+        which can be used to quantify a choice of forecaster for the time series.
+
+    fallback_forecaster : sktime forecaster | None, Optional
+        A fallback forecaster that will be used if the category generated by
+        the transformer does not match any of the given forecasters.
+
+    Raises
+    ------
+    AssertionError: If a valid transformer (an instance of BaseTransformer)
+    is not passed or if valid forecasters (instances of BaseForecaster) are not given.
+
+    Examples
+    --------
+    >>> from sktime.forecasting.compose._category_compositor import CategoryCompositor
+    >>> from sktime.forecasting.croston import Croston
+    >>> from sktime.forecasting.trend import PolynomialTrendForecaster
+    >>> from sktime.forecasting.naive import NaiveForecaster
+    >>> from sktime.transformations.series.adi_cv import ADICVTransformer
+    >>> from sktime.transformations.series.tests.test_adi_cv import (
+    ...     _generate_erratic_series)
+    >>> category_compositor = CategoryCompositor(
+    ...     forecasters =
+    ...         {"smooth": NaiveForecaster(),
+    ...         "erratic": Croston(),
+    ...         "intermittent": PolynomialTrendForecaster()},
+    ...     transformer=ADICVTransformer(features=["class"]))
+
+    >>> generated_data = _generate_erratic_series()
+    >>> category_compositor = category_compositor.fit(generated_data, fh=50)
+
+    >>> # Verifies that the correct forecaster is chosen
+    >>> # print(category_compositor.chosen_forecaster_)
+
+    >>> # Print out the predicted value!
+    >>> # print(category_compositor.predict(fh=50, X=None))
+    """
+
+    _tags = {
+        "y_inner_mtype": "pd.DataFrame",
+        "X_inner_mtype": "pd.DataFrame",
+        "scitype:y": "both",
+        "ignores-exogeneous-X": False,
+        "requires-fh-in-fit": False,
+        "enforce_index_type": None,
+        "authors": ["shlok191"],
+        "maintainers": ["shlok191"],
+        "python_version": None,
+    }
+
+    def __init__(
+        self,
+        forecasters,
+        transformer=adi_cv_transformer,
```

Comment:
mutable objects should never be used as defaults in python methods or classes!

The reason is, if the object is mutated, then the default for any future calls is also mutated.

The pythonic way to do is is to have the defalut = `None`, and then check if is `None`, then set the param to the default.

Now, in `sklearn` API, params should be written to self and never overwritten, so you ought to write the default or transformer to `self.transformer`.

Reply:
Ah, got it. That makes sense. I'll ammend this and add it to the next commit.